elastic/kibana
# [Index Management] Add support for effective retention in DSL

## Summary
The ES team has introduced the term factory retention to capture the ability to set the default and max retention at startup via file-based settings. This alleviates the need to provide feedback to the user about the impact of changing these values.

This PR surfaces the effective retention value to the user so that its clearer for them which data retention value takes precedence.

##### How to test
* Start ES with `node scripts/es serverless --tag latest --projectType oblt -E data_streams.lifecycle.retention.factory_default=30d -E data_streams.lifecycle.retention.factory_max=30d` and kibana with `yarn serverless-oblt`
* Navigate to console and create a data stream following [this instructions](https://www.elastic.co/guide/en/elasticsearch/reference/current/set-up-a-data-stream.html#create-component-templates)
* Navigate to `Index Management` -> `Data streams`
* Click on the newly created DS and verify that changing the data retention to be `33d` throws a warning
* Verify that the table highlights the DS given that now its using an effective retention !== than the user defined one


<details>
<summary>Screenshots</summary>

// Table tooltip
<img width="1246" alt="Screenshot 2024-05-01 at 10 30 42" src="https://github.com/elastic/kibana/assets/1191206/3d40fcd5-685e-4cf0-99fc-9bb71a4e3fba">

// Tooltip with es warning when the user retention saved by the user wont be used
![Screenshot 2024-05-02 at 13 45 48](https://github.com/elastic/kibana/assets/1191206/6d6725aa-de8a-4dc3-8128-a449252737ae)

// Flyout with DS details
<img width="493" alt="Screenshot 2024-05-01 at 10 26 04" src="https://github.com/elastic/kibana/assets/1191206/560d054e-94f2-471c-89e0-eb9b01e3d61f">

<img width="492" alt="Screenshot 2024-05-01 at 10 26 39" src="https://github.com/elastic/kibana/assets/1191206/00717ca4-0bf5-4a88-80c7-8f2886d2f1e7">

</details>

<details>
<summary>Testing mocks</summary>

At the time of writing, the required ES changes havent been incorporated into the latest snapshot for the docker image. So in order to test effective_data retention with a max retention set grab the following [mock.txt](https://github.com/elastic/kibana/files/15187839/mock.txt) and apply it with `git apply ~/path/mock.txt` into the root of the kibana repo. Make sure to update the DS name you want to use to get the mock applied.


</details>
